File: Binary Search/6. HENGAP/Hengap.py
import sys, bisect
sys.stdin = open('Hengap.inp', 'r')
sys.stdout = open('Hengap.out', 'w')

n, r = map(int, input().split())
a = list(map(int, input().split()))
a.sort()

res = 0
for i in range(n):
    res += n - bisect.bisect_left(a, a[i] + r + 1, i+1)
print(res)

File: Binary Search/6. HENGAP/Hengap.cpp
#include <bits/stdc++.h>
using namespace std;

long long n, a[300001], r, res, mid;

int main() {
    ios_base::sync_with_stdio(0); cin.tie(0); cout.tie(0);
    freopen("Hengap.inp", "r", stdin);
    freopen("Hengap.out", "w", stdout);

    cin >> n >> r;

    for (int i = 0; i < n; i++) cin >> a[i];
    sort(a, a + n);

    for (int i = 0; i < n; i++) res += n - (lower_bound(a, a+n, a[i] + r + 1) - a);
    
    cout << res << '\n';
}